Yosemite/Yosemite/Stores/ShippingLabelStore.swift

```diff
@@ -36,17 +36,109 @@ public final class ShippingLabelStore: Store {
+        derivedStorage.perform { [weak self] in
+            guard let self = self else { return }
+            guard let order = derivedStorage.loadOrder(siteID: siteID, orderID: orderID) else {
+                return
+            }
+            guard shippingLabels.isEmpty == false else {
+                return
+            }
+            self.upsertShippingLabels(siteID: siteID, orderID: orderID, shippingLabels: shippingLabels, storageOrder: order)
+            self.upsertShippingLabelSettings(siteID: siteID, orderID: orderID, settings: settings, storageOrder: order)
+        }
+
+        storageManager.saveDerivedType(derivedStorage: derivedStorage) {
+            DispatchQueue.main.async(execute: onCompletion)
+        }
+    }
+
+    /// Updates/inserts the specified readonly ShippingLabel entities in the current thread.
+    func upsertShippingLabels(siteID: Int64, orderID: Int64, shippingLabels: [ShippingLabel], storageOrder: StorageOrder) {
+        let derivedStorage = sharedDerivedStorage
+
+        for shippingLabel in shippingLabels {
+            let storageShippingLabel = derivedStorage.loadShippingLabel(siteID: shippingLabel.siteID,
+                                                                        orderID: shippingLabel.orderID,
+                                                                        shippingLabelID: shippingLabel.shippingLabelID) ??
+                derivedStorage.insertNewObject(ofType: Storage.ShippingLabel.self)
+            storageShippingLabel.update(with: shippingLabel)
+            storageShippingLabel.order = storageOrder
+
+            if let refund = shippingLabel.refund {
+                let storageRefund = storageShippingLabel.refund ?? derivedStorage.insertNewObject(ofType: Storage.ShippingLabelRefund.self)
+                storageRefund.update(with: refund)
+                storageShippingLabel.refund = storageRefund
+            }
+
+            let originAddress = storageShippingLabel.originAddress ?? derivedStorage.insertNewObject(ofType: Storage.ShippingLabelAddress.self)
+            originAddress.update(with: shippingLabel.originAddress)
+            storageShippingLabel.originAddress = originAddress
+
+            let destinationAddress = storageShippingLabel.destinationAddress ?? derivedStorage.insertNewObject(ofType: Storage.ShippingLabelAddress.self)
+            destinationAddress.update(with: shippingLabel.destinationAddress)
+            storageShippingLabel.destinationAddress = destinationAddress
+        }
+
+        // Now, remove any objects that exist in storage but not in shippingLabels
+        let storedShippingLabels = derivedStorage.loadAllShippingLabels(siteID: siteID, orderID: orderID)
+        storedShippingLabels.forEach { storedShippingLabel in
+            if !shippingLabels.contains(where: { $0.shippingLabelID == storedShippingLabel.shippingLabelID }) {
+                derivedStorage.deleteObject(storedShippingLabel)
+            }
+        }
```

Comment:
A `filter` may make this more readable. Maybe. 🙃 

```suggestion
        let shippingLabelIDs = shippingLabels.map(\.shippingLabelID)
        derivedStorage.loadAllShippingLabels(siteID: siteID, orderID: orderID).filter {
            !shippingLabelIDs.contains($0.shippingLabelID)
        }.forEach {
            derivedStorage.deleteObject($0)
        }
```

Reply:
Nice this looks cleaner! updated here: https://github.com/woocommerce/woocommerce-ios/pull/3176/commits/bf585bf160a3671e65a51350e1a255f5350b740e